### ISSUE TITLE
Wrong inferred type param when passing `this` to Microsoft.JSInterop.DotNetObjectRef

### ISSUE THREAD

**Issue by chucker** [aspnet/AspNetCore#12446]
Preview 6 replaces `DotNetObjectRef` with a generic `DotNetObjectRef<T>`. Instead of a constructor, the `Create<T>` method is now used.

When passing `this` from within a component, this leads to unexpected results due to Razor's parsing behavior.

### To Reproduce
1. Using ASP.NET Core **3.0 Preview 6** and Visual Studio 2019 **16.2 Preview 4**
2. Create a new Blazor component **MySampleComponent.razor**
3. Place this code inside its `@code` section:

```csharp
    private DotNetObjectRef<MySampleComponent> _Instance;

    protected override async Task OnAfterRenderAsync()
    {
        _Instance = DotNetObjectRef.Create(this);
    }
```

### Expected behavior
This should compile. The type parameter of `Create<T>` should be inferred as `MySampleComponent`.

### Actual behavior
This doesn't work, because Razor internally renames your class to `__generated__MySampleComponent`. Consequently, the Error List will show:

> CS0029 Cannot implicitly convert type `Microsoft.JSInterop.DotNetObjectRef<__generated__MySampleComponent>` to `Microsoft.JSInterop.DotNetObjectRef<MySampleComponent>`

If you explicitly specify the type parameter, the error changes to CS1503.

### Workaround
You can declare the variable with the generated class name:

```csharp
    private DotNetObjectRef<__generated__MySampleComponent> _Instance;

    protected override async Task OnAfterRenderAsync()
    {
        _Instance = DotNetObjectRef.Create(this);
    }
```

However, this is presumably not the intended behavior.

**Comment by chucker** [comment]
I filed this a bit prematurely. The above error messages are real. However, when actually building the project, _you need to go with the original code_.

So, while this will show an error during Razor compilation:

```csharp
    private DotNetObjectRef<MySampleComponent> _Instance;

    protected override async Task OnAfterRenderAsync()
    {
        _Instance = DotNetObjectRef.Create(this);
    }
```

, it is in fact the correct syntax for building the project.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @chucker.
This will be resolved as part of #4065
